switching this lesson to the new Carpentries Workbench Template (#23)
hpc-carpentry/hpc-chapel: episodes/02-variables.md

---
title: "Basic syntax and variables"
teaching: 15
exercises: 15
---

:::::::::::::::::::::::::::::::::::::: questions
- "How do I write basic Chapel code?"
::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives
- "Perform basic maths in Chapel."
- "Understand Chapel's basic data types."
- "Understand how to read and fix errors."
- "Know how to define and use data stored as variables."
::::::::::::::::::::::::::::::::::::::::::::::::

> Using basic maths in Chapel is fairly intuitive. Try compiling the following code to see

Reviewer: Contextualize this for beginners, like the Ep. 1 examples:
Open a text editor and populate the document with these lines; save it as `operators.chpl`; compile and run with the following command.

how the different mathematical operators work.

```chpl
writeln(4 + 5);
writeln(4 - 5);
writeln(4 * 5);
writeln(4 / 5);   // integer division
writeln(4.0 / 5.0); // floating-point division
writeln(4 ** 5);  // exponentiation
```

In this example, our code is called `operators.chpl`. You can compile it with the following commands:

```bash
chpl operators.chpl --fast
./operators.o
```

You should see output that looks something like the following:

```output
9
-1
20
0
0.8
1024
```

Code beginning with `//` is interpreted as a comment &mdash; it does not get run. Comments are very valuable
when writing code, because they allow us to write notes to ourselves about what each piece of code does. You
can also create block comments with `/*` and `*/`:

```chpl
/* This is a block comment.
It can span as many lines as you want!
(like this) */
```

> 

Reviewer: Better practice would make that edit, or create a block comment describing what the code does in `operators.chpl`, then recompile & run. This will catch a few learners' typos.

it. Although we already kind of know why the error was caused (we tried to reassign the value of a `const`
variable, which by definition cannot be changed), let's walk through the error as an example of how to
troubleshoot our programs.

- `variables.chpl:2:` indicates that the error was caused on line 2 of our `variables.chpl` file.

- `error:` indicates that the issue was an error, and blocks compilation.  Sometimes the compiler will just
  give us warning or information, not necessarily errors. When we see something that is not an error, we
  should carefully read the output and consider if it necessitates changing our code.  Errors must be fixed,
  as they will block the code from compiling.

- `cannot assign to const variable` indicates that we were trying to reassign a `const` variable, which is
  explicitly not allowed in Chapel.

To fix this error, we can change `const` to `var` when declaring our `test` variable. `var` indicates a
variable that can be reassigned.

```chpl
var test = 100;
test = 200;
writeln('The value of test is: ', test);
writeln(test / 4);
```
```bash
chpl variables.chpl
```
```output
The value of test is: 200
50
```

> 

Reviewer: Too many blank lines.

In Chapel, to initialize a variable we must specify the type of the variable, or initialise it in place with
some value. The common variable types in Chapel are:

- integer `int` (positive or negative whole numbers)
- floating-point number `real` (decimal values)
- Boolean `bool`  (true or false)
- string `string` (any type of text)

These two variables below are initialized with the type. If no initial value is given, Chapel will initialise
a variable with a default value depending on the declared type, for example 0 for integers and 0.0 for real
variables.

```chpl
var counter: int;
var delta: real;
writeln("counter is ", counter, " and delta is ", delta);
```
```bash
chpl variables.chpl
./variables.o
```
```output
counter is 0 and delta is 0.0
```

If a variable is initialised with a value but without a type, Chapel will infer its type from the given
initial value:

```chpl
const test = 100;

> writeln('The value of test is ', test, ' and its type is ', test.type:string);

Reviewer: Neato, and please talk through the syntax of `test.type:string`

title: "Basic syntax and variables"
teaching: 15
exercises: 15
---

:::::::::::::::::::::::::::::::::::::: questions
- "How do I write basic Chapel code?"
::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives
- "Perform basic maths in Chapel."
- "Understand Chapel's basic data types."
- "Understand how to read and fix errors."
- "Know how to define and use data stored as variables."
::::::::::::::::::::::::::::::::::::::::::::::::

> Basic maths in Chapel works the same as other programming languages. Try compiling the following code to see

Reviewer: "... Chapel works the same as other programming languages" is not a meaningful statement: there is no universal syntax for arithmetic, as the provided examples show. Do you mean it works like C/C++/Fortran, or something about how Chapel implicitly types numerical literals based on the presence of specific characters (`1`, `1.`, `0+1i`)? Clarify for the learners, please.

Author: Perhaps, the writer meant that basic maths in Chapel are fairly intuitive? I changed the text to say that.

Author: Marking this as resolved through a commit.

how the different mathematical operators work.

```chpl
writeln(4 + 5);
writeln(4 - 5);
writeln(4 * 5);
writeln(4 / 5);   // integer division

> writeln(4.0 / 5.0); // normal division

Reviewer: "Normal division" doesn't exist, this is "floating-point division".

Author: Resolved through a commit.

writeln(4 ** 5);  // exponentiation
```

In this example, our code is called `operators.chpl`. You can compile it with the following commands:

```bash
chpl operators.chpl --fast -o operators.o
./operators.o
```

You should see output that looks something like the following:

```output
9
-1
20
0
0.8
1024
```

Code beginning with `//` is interpreted as a comment &mdash; it does not get run. Comments are very valuable
when writing code, because they allow us to write notes to ourselves about what each piece of code does. You
can also create block comments with `/*` and `*/`:

```chpl
/* This is a block comment.
It can span as many lines as you want!
(like this) */
```

## Variables

Granted, we probably want to do more than basic maths with Chapel. We will need to store the results of
complex operations using variables. Variables in programming are not the same as the mathematical concept. In

> programming, a variable is an allocated space in the memory of the computer, where we can store information or

Reviewer: "a variable is an allocated space in the memory of the computer"---unsure this is a helpful statement. Heap memory is explicitly allocated in the code; stack memory is pre-allocated; the difference is likely beyond the scope of the lesson.

Consider something like, "a _variable_ represents (or references) a location in the memory of the computer where we can store information or..."

Author: Suggestion accepted. Resolved through a commit.

data while executing a program. A variable has three elements:

1. a **_name_** or label, to identify the variable 
2. a **_type_**, that indicates the kind of data that we can store in it, and
3. a **_value_**, the actual information or data stored in the variable.

When we store a value in a variable for the first time, we say that we **_initialised_** it. Further changes

> to the value of a variable are called **_assignments_**, in general, `x=a` means that we assign the value *a*

Reviewer: Imprecise language: does `a` have a value?

Author: I removed that entire paragraph. We talk more about this distinction later in this chapter. Resolved through a commit.

to the variable *x*.

Variables in Chapel are declared with the `var` or `const` keywords. When a variable declared as const is

> initialised, its value cannot be modified anymore during the execution of the program.

Reviewer: Example does not fully align with explanation, since `const test = 100;` represents simultaneous _declaration and initialization_. The following more closely aligns with the text: does it compile?

```chpl
const test;  // declare 'test' variable
test = 100;  // initialize 'test' with specified value
writeln('The value of test is: ', test);
```

If not, please rephrase for clarity.

Reviewer: (It does indeed compile. This may be a useful "discussion" call-out.)

Reviewer: (Also: the compiled size is 2 MB! Yikes!)

Author: I added a call-out. Resolved through a commit.

initial value. We can use the stored variable simply by using its name anywhere
in our code (called `variables.chpl`).

```chpl

> const test = 100;

Reviewer: The body introduces `int` and `real` as "common variable types," but the example is implicit about type. Consider refactoring to convey `const` vs. `var` first, using implicitly-cast integers, then introduce the types (`string` should also be mentioned) when those statically-cast assignments are made later in the lesson material.

Author: Done. I rewrote 02-variables.md changing the order in which these concepts are introduced. Resolved through a commit.

writeln('The value of test is: ', test);
writeln(test / 4);
```

```bash
chpl variables.chpl -o variables.o --fast
./variables.o
```

```output
The value of test is: 100
25
```

This constant variable `test` will be created as an integer, and initialised with the value 100. No other
values can be assigned to these variables during the execution of the program. What happens if we try to
modify a constant variable like `test`?

```chpl
const test = 100;
test = 200;
writeln('The value of test is: ', test);
writeln(test / 4);
```

```bash
chpl variables.chpl -o variables.o
```

```error
variables.chpl:2: error: illegal lvalue in assignment
```

The compiler threw an error, and did not compile our program. This is a feature of compiled languages - if

> there is something wrong, we will typically see an error while writing our program, instead of while running

Reviewer: "while writing" -> "while building" or "at compile-time"

Author: Suggestion accepted. Resolved through a commit.

* `variables.chpl:2:` indicates that the error was caused on line 2 of our `variables.chpl` file.

* `error:` indicates that the issue was an error, and blocks compilation.  Sometimes the compiler will just
  give us warning or information, not necessarily errors. When we see something that is not an error, we
  should carefully read the output and consider if it necessitates changing our code.  Errors must be fixed,
  as they will block the code from compiling.

* `illegal lvalue in assignment` indicates that the left hand side of our assignment expression (`lvalue`) was
  illegal. We were trying to reassign a `const` variable, which is explicitly not allowed in Chapel.

To fix this error, we can change `const` to `var` when declaring our `test` variable. `var` indicates a
variable that can be reassigned.

```chpl
var test = 100;
test = 200;
writeln('The value of test is: ', test);
writeln(test / 4);
```

```bash
chpl variables.chpl -o variables.o
```

```output
The value of test is: 200
50
```

It worked! Now we know both how to set, use, and change a variable, as well as the implications of using `var`
and `const`. We also know how to read and interpret errors.

> ## Uninitialised variables

Reviewer: This section introduces temperature-themed variable names, without any explanation or motivation. Insert a story or example physical system for context, or revise the code to use fictitious variable names that don't require any context.

Author: Done. I added three paragraphs referring to the previous section and describing all variables before introducing them in the code. Resolved through a commit.

On the other hand, if a variable is declared without an initial value, Chapel will initialise it with a
default value depending on the declared type (0.0 for real variables, for example). The following variables
will be created as real floating point numbers equal to 0.0.

```chpl

> var curdif: real;	//here we will store the greatest difference in temperature from one iteration to another 

Reviewer: Introducing _iterations_ here is a distracting mistake since the code does not include any looping construct.

Author: I disagree. We introduced the problem in 01-intro.md, and there we talk about iterations. Here we are introducing variables that we'll use later in the code with loops.

> var tt: real;		//for temporary results when computing the temperatures

Reviewer: `tt` is a terse variable name. So are `curdif`, `mindif`, and `n`. Consider teaching verbose variable naming conventions, using underscores, etc.: this ain't FORTRAN 77!

Author: These names are very subjective. In my original version `tt` was called `tmp` (it's a temporary variable), `curdif` was called `delta` (it is the temperature difference), `mindif` was called `tolerance`, and so on. Jeff S. edited these files in 2017 and changed these names to what you see now. I will simply change them back to what they were in my version.

Author: Changed the variable names back to their original names.

```

> Of course, we can use both, the initial value and the type, when declaring a variable as follows:

Reviewer: Don't say "of course." Be clear about declaration, initialization, and assignment.

```suggestion
When declaring a variable, we can assign its type in addition to its initial value:
```

Author: Suggestion accepted. Also corrected the syntax (type comes before value) and added a call-out to explain. Resolved through a commit.

const mindif = 0.0001: real;    // smallest difference in temperature that would be accepted before stopping
const n = 20: int;              // the temperature at the desired position will be printed every n iterations
```

::::::::::::::::::::::::::::::::::::: keypoints
- "A comment is preceded with `//` or surrounded by `/* and `*/`"

> - "All variables hold a certain type of data."

Reviewer: This statement is false:

> "All variables hold a certain type of data."

From the [User's Guide](https://chapel-lang.org/docs/2.1/users-guide/base/vars.html), "a variable declaration may omit its initial value or its type, but not both." -- and when the type is not specified, it will be implicitly cast in a type that is capable of representing the value. This is not a "certain" type.

Author: Not sure I understand the objection. After a variable is cast in a type that is capable of representing the value, it will have a type. The point here is that all variables have a type.

Reviewer: The User Guide is careful about variable initialization. It is true and important to convey that "variables are typed," but the specific statement that "all variables hold a _certain_ type" is misleading. The Chapel compiler will choose a type based on the value, if a type is not provided; but this implicitly-cast variable may be of _any_ type that can represent the value. This is not "certain," it's probably implementation-dependent.

Author: Ok, I reformulated the key point -- will the new version work?

> - "Using `const` instead of `var` prevents reassignment."

Reviewer: This statement is also false:

> "Using `const` instead of `var` prevents reassignment."

Multiple assignments to a `const` variable prevents the compiler from successfully building the program. This prevents reassignment in the broadest possible sense, but clarity would be preferable.

Author: I reworded that key point. Resolved through a commit.
